osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/SetEncryptionKeyExchangeOnGMeterCommandExecutor.java

```diff
@@ -82,77 +85,79 @@ public MethodResultCode execute(final DlmsConnectionManager conn, final DlmsDevi
-            methodResultCode = conn.getConnection().action(methodSetEncryptionKey);
-            this.checkMethodResultCode(methodResultCode, "M-Bus Setup set_encryption_key");
-            LOGGER.info("Successfully invoked M-Bus Setup set_encryption_key method: class_id {} obis_code {}",
-                    CLASS_ID, obisCode);
-
-            this.securityKeyService.validateNewKey(mbusDevice, mbusUserKeyType);
+            methodResultCode = this.setEncryptionKey(conn, channel, encryptionKey);
+            this.checkMethodResultCode(methodResultCode, "M-Bus Setup set_encryption_key", obisCode);
 
+            this.securityKeyService.activateNewKey(mbusDeviceIdentification, G_METER_ENCRYPTION);
             return MethodResultCode.SUCCESS;
         } catch (final IOException e) {
             throw new ConnectionException(e);
         } catch (final EncrypterException e) {
             throw new ProtocolAdapterException(
                     "Unexpected exception during decryption of security keys, reason = " + e.getMessage(), e);
+        } catch (final FunctionalException e) {
+            throw new ProtocolAdapterException("Unexpected exception: " + e.toString(), e);
         }
     }
 
-    private void checkMethodResultCode(final MethodResult methodResultCode, final String methodParameterName)
-            throws ProtocolAdapterException {
+    private MethodResult setEncryptionKey(DlmsConnectionManager conn, int channel, byte[] encryptionKey)
+            throws IOException {
+        MethodResult methodResultCode;
+        final MethodParameter methodSetEncryptionKey = this
+                .getSetEncryptionKeyMethodParameter(OBIS_HASHMAP.get(channel), encryptionKey);
+        conn.getDlmsMessageListener().setDescription("SetEncryptionKeyExchangeOnGMeter for channel " + channel
+                + ", call M-Bus Setup set_encryption_key method: " + JdlmsObjectToStringUtil
+                .describeMethod(methodSetEncryptionKey));
+        methodResultCode = conn.getConnection().action(methodSetEncryptionKey);
+        return methodResultCode;
+    }
+
+    private MethodResult transferKey(DlmsConnectionManager conn, String mbusDeviceIdentification, int channel,
+            byte[] encryptionKey) throws ProtocolAdapterException, IOException, FunctionalException {
+        final MethodParameter methodTransferKey = this
+                .getTransferKeyMethodParameter(mbusDeviceIdentification, channel, encryptionKey);
+        conn.getDlmsMessageListener().setDescription(
+                "SetEncryptionKeyExchangeOnGMeter for channel " + channel + ", call M-Bus Setup transfer_key method: "
+                        + JdlmsObjectToStringUtil.describeMethod(methodTransferKey));
+
+        return conn.getConnection().action(methodTransferKey);
+    }
+
+    private MethodParameter getTransferKeyMethodParameter(String mbusDeviceIdentification, int channel,
+            byte[] encryptionKey) throws ProtocolAdapterException, FunctionalException {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "encryptionKey". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1172)